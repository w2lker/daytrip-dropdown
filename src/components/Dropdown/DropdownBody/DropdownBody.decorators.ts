import withStyles from "react-jss";

import DropdownBody from "./DropdownBody";
import dropdownBodyStyles from "./DropdownBody.styles";

const styledComponent = withStyles(dropdownBodyStyles)(DropdownBody);

export default styledComponent
