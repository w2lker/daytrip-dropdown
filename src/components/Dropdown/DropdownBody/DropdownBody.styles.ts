const dropdownBodyStyles = {
    wrapper: {
      backgroundColor: '#ECF0F2',
      borderRadius: 4,
      boxShadow: '0 2px 5px 0 rgba(146,166,176,0.50)',
    },

    contentWrapper: {
      overflowX: 'hidden',
      overflowY: 'scroll',
      maxHeight: 'calc(100vh - 300px)',
      position: 'relative'
    },
};

export default dropdownBodyStyles;
