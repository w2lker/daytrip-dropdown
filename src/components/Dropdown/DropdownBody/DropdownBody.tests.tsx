describe('DropdownBody', () => {
    test.todo('Placeholder');
    test.todo('match snapshot');
});

describe('DropdownBody styles', () => {
    test.todo('has principle classes');
});

describe('DropdownBody decorated', () => {
    test.todo('provide classes prop to decorated component');
});
