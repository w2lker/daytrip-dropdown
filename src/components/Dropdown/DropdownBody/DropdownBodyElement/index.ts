import DropdownBodyElement from './DropdownBodyElement.decorators';

export default DropdownBodyElement;
