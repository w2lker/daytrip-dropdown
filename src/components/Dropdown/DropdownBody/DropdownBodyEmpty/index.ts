import DropdownBodyEmpty from './DropdownBodyEmpty.decorators';

export default DropdownBodyEmpty;
