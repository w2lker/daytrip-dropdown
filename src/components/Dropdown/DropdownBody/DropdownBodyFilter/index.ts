import DropdownBodyFilter from './DropdownBodyFilter.decorators';

export default DropdownBodyFilter;
