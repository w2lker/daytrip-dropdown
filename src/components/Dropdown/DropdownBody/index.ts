import DropdownBody from "./DropdownBody.decorators";

export default DropdownBody;
