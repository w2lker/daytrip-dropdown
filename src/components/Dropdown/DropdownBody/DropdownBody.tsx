import React, { useState } from "react";

import dropdownBodyStyles from "./DropdownBody.styles";
import DropdownBodyElement from "./DropdownBodyElement";
import DropdownBodyFilter from "./DropdownBodyFilter";
import DropdownBodyEmpty from "./DropdownBodyEmpty";

import {IDropdownOptionsArray} from "../DropdownRoot/DropdownRoot";
import {getEntries, getKey, getValue} from "../../../utils/dropdown";

export interface IDropdownBodyProps {
    opened: boolean;
    selected?: string;
    options: IDropdownOptionsArray;
    rows?: number;
    caseInsensitiveSearch?: boolean;
    onSelect: (val: string) => void;
    onClose: () => void;
}

const filterOptions = (filter: string, options: IDropdownOptionsArray, caseInsensitive?: boolean): IDropdownOptionsArray => {
  const insensitive = caseInsensitive ? 'i' : '';
  const regExpFilter = new RegExp(filter, insensitive);
  return options.filter( (opt) => getValue(opt).match(regExpFilter));
};


const DropdownBody: React.FC<IDropdownBodyProps> = (props) => {
    const { opened, selected, options, onSelect, onClose, caseInsensitiveSearch } = props;
    const [filter, setFilter] = useState('');
    const [focused, setFocused] = useState('');

    if (!opened) {
      return null;
    }

    const filtered = filterOptions(filter, options, caseInsensitiveSearch);
    const classes = dropdownBodyStyles(props);

    const focusSwitch = (next: boolean) => () => {
      if (!filtered) return;
      const setNextFocused = (index: number) => {
        const next = getKey(filtered[index]);
        setFocused(next);
      };
      const index = filtered.findIndex( (item) =>  getKey(item) === focused);
      const filteredLast = filtered.length - 1;
      // Can't select next to last and prev to first element
      if (index === filteredLast && next) return;
      if (index === 0 && !next) return;
      // If focused doesn't match filtered list
      if (index === -1) {
        return setNextFocused(0);
      }
      setNextFocused(next ? index + 1 : index - 1);
    };

    const renderElements = filtered.map( (item) => {
      const {key, value: content} = getEntries(item);
      const handleSelect = () => {
        onSelect(key);
        onClose();
      };

      return (
        <DropdownBodyElement
          key={key}
          content={content}
          selected={ key === selected }
          focused={ key === focused }
          onClick={ handleSelect }
        />
      )
    });

    return (
        <div className={classes.wrapper}>
          <DropdownBodyFilter
            value={filter}
            onClose={onClose}
            onFilterChange={setFilter}
            onSelectPerform={ () => focused && onSelect(focused) }
            onSelectNext={ focusSwitch(true) }
            onSelectPrev={ focusSwitch(false) }
          />
          <div className={classes.contentWrapper}>
            {renderElements}
            <DropdownBodyEmpty
              isEmpty={!(options && options.length)}
              isFilteredEmpty={!(filtered && filtered.length)}
            />
          </div>
        </div>
    )
};

export default DropdownBody;
