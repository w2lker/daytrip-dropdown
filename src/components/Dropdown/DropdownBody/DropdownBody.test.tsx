import React from 'react';

import { mount } from 'enzyme';

import DropdownBody, { IDropdownBodyProps } from './DropdownBody';
import DropdownBodyDecorated from './DropdownBody.decorators';
import dropdownBodyStyles from './DropdownBody.styles';
import {getKey, getValue} from "../../../utils/dropdown";

const setup = () => {
  const classes = {
    wrapper: 'dropdownBodyFilter-wrapper',
    contentWrapper: 'dropdownBodyFilter-input',
  };

  const classesKeys = Object.keys(classes);
  const sampleProps: IDropdownBodyProps = {
    opened: true,
    selected: 'val1',
    options: [{'val1': 'content1'}, {'val2': 'content2'}],
    onSelect: jest.fn(),
    onClose: jest.fn(),
    classes,
  };
  const componentContent = {
    own: 'DropdownBody',
    filter: 'DropdownBodyFilter',
    filterInput: 'DropdownBodyFilter input',
    element: 'DropdownBodyElement',
  };
  return {classes, classesKeys, sampleProps, componentContent}
};

describe('DropdownBody component', () => {

  it('match snapshot', () => {
    const { sampleProps } = setup();
    const component = mount(<DropdownBody {...sampleProps} />);
    expect(component.debug()).toMatchSnapshot();
  });

  it('renders unfiltered list', () => {
      const { sampleProps, componentContent } = setup();
      const component = mount(<DropdownBody {...sampleProps} />);
      const list = component.find(componentContent.element);
      expect(list.length).toBe(2);
      list.forEach( (node, index) => {
        expect(node.props().content).toBe(getValue(sampleProps.options[index]));
      })
  });

  it('filters content', () => {
    const { sampleProps, componentContent } = setup();
    const component = mount(<DropdownBody {...sampleProps} />);
    const filterInput = component.find(componentContent.filterInput);
    const filterValue = getValue(sampleProps.options[0]);
    filterInput.simulate('change', {target: {value: filterValue}});
    const list = component.find(componentContent.element);
    expect(list.length).toBe(1);
    expect(list.get(0).props.content).toMatch(filterValue);
  });

  it('focus element', () => {
    const { sampleProps, componentContent } = setup();
    const component = mount(<DropdownBody {...sampleProps} />);
    const filterInput = component.find(componentContent.filterInput);
    filterInput.simulate('keyDown', {
      key: 'Tab',
      keyCode: 9,
    });
    expect(component.find(componentContent.element).get(0).props.focused).toBeTruthy();
  });

  it('select element', () => {
    const { sampleProps, componentContent } = setup();
    const component = mount(<DropdownBody {...sampleProps} />);
    component.find(`${componentContent.element} div`).first().simulate('mouseDown');
    // @ts-ignore
    const selectedKey = sampleProps.onSelect.mock.calls[0][0];
    expect(selectedKey).toBe(getKey(sampleProps.options[0]));
  });

  it('perform onClose on element selection', () => {
    const { sampleProps, componentContent } = setup();
    const component = mount(<DropdownBody {...sampleProps} />);
    component.find(`${componentContent.element} div`).first().simulate('mouseDown');
    expect(sampleProps.onClose).toBeCalledTimes(1);
  });

  it('perform onClose on filter blur',() => {
    const { sampleProps, componentContent } = setup();
    const component = mount(<DropdownBody {...sampleProps} />);
    const filterInput = component.find(componentContent.filterInput);
    filterInput.simulate('blur' );
    expect(sampleProps.onClose).toBeCalledTimes(1);
  })

});

describe('DropdownBody styles', () => {
  it('contains principle classes', () => {
    const { classesKeys } = setup();
    expect(dropdownBodyStyles).toBeDefined();
    classesKeys.forEach((keyValue) => {
      // @ts-ignore
      expect(dropdownBodyStyles[keyValue]).toBeDefined();
    });
  });
});

describe('DropdownBody decorators', () => {
  it('provides styled classes from decorators', () => {
    const { classesKeys, sampleProps } = setup();
    const component = mount(<DropdownBodyDecorated {...sampleProps} />);
    // @ts-ignore
    const assignedClasses = component.find('DropdownBody').props().classes;
    classesKeys.forEach(keyValue => {
      expect(assignedClasses[keyValue]).toBeDefined();
    });
  });
});
