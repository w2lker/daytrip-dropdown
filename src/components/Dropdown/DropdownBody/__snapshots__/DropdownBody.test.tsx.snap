// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownBody component match snapshot 1`] = `
"<DropdownBody opened={true} selected=\\"val1\\" options={{...}} onSelect={[Function: mockConstructor]} onClose={[Function: mockConstructor]} classes={{...}}>
  <div className=\\"wrapper-0-2-1\\">
    <JssContextSubscriber value=\\"\\" onClose={[Function: mockConstructor]} onFilterChange={[Function: bound dispatchAction]} onSelectPerform={[Function: onSelectPerform]} onSelectNext={[Function]} onSelectPrev={[Function]}>
      <WithStyles(DropdownBodyFilter) innerRef={{...}} value=\\"\\" onClose={[Function: mockConstructor]} onFilterChange={[Function: bound dispatchAction]} onSelectPerform={[Function: onSelectPerform]} onSelectNext={[Function]} onSelectPrev={[Function]} jssContext={{...}} theme={{...}}>
        <DropdownBodyFilter value=\\"\\" onClose={[Function: mockConstructor]} onFilterChange={[Function: bound dispatchAction]} onSelectPerform={[Function: onSelectPerform]} onSelectNext={[Function]} onSelectPrev={[Function]} classes={{...}}>
          <div className=\\"DropdownBodyFilter-wrapper-0-2-4\\">
            <div className=\\"DropdownBodyFilter-magnify-0-2-5\\" onClick={[Function: performFocus]}>
              <svg width=\\"14px\\" height=\\"14px\\" viewBox=\\"0 0 14 14\\">
                <path d=\\"M5.2,0 C8.072,0 10.4,2.328 10.4,5.2 C10.4,6.488 9.928,7.672 9.152,8.584 L9.368,8.8 L10,8.8 L14,12.8 L12.8,14 L8.8,10 L8.8,9.368 L8.584,9.152 C7.672,9.928 6.488,10.4 5.2,10.4 C2.328,10.4 0,8.072 0,5.2 C0,2.328 2.328,0 5.2,0 Z M5.2,1.6 C3.2,1.6 1.6,3.2 1.6,5.2 C1.6,7.2 3.2,8.8 5.2,8.8 C7.2,8.8 8.8,7.2 8.8,5.2 C8.8,3.2 7.2,1.6 5.2,1.6 Z\\" fill=\\"#299BF6\\" />
              </svg>
            </div>
            <input className=\\"DropdownBodyFilter-input-0-2-6\\" value=\\"\\" placeholder=\\"\\" onChange={[Function: handleFilterChange]} onKeyDown={[Function: handleFilterKeyDown]} />
            <div className=\\"DropdownBodyFilter-divider-0-2-7\\" />
          </div>
        </DropdownBodyFilter>
      </WithStyles(DropdownBodyFilter)>
    </JssContextSubscriber>
    <div className=\\"contentWrapper-0-2-2 contentWrapper-0-0-2-3\\">
      <JssContextSubscriber content=\\"content1\\" selected={true} focused={false} onClick={[Function: handleSelect]}>
        <WithStyles(DropdownBodyElement) innerRef={{...}} content=\\"content1\\" selected={true} focused={false} onClick={[Function: handleSelect]} jssContext={{...}} theme={{...}}>
          <DropdownBodyElement content=\\"content1\\" selected={true} focused={false} onClick={[Function: handleSelect]} classes={{...}}>
            <div className=\\"DropdownBodyElement-root-0-2-8 DropdownBodyElement-selected-0-2-10\\" onClick={[Function: handleSelect]}>
              content1
            </div>
          </DropdownBodyElement>
        </WithStyles(DropdownBodyElement)>
      </JssContextSubscriber>
      <JssContextSubscriber content=\\"content2\\" selected={false} focused={false} onClick={[Function: handleSelect]}>
        <WithStyles(DropdownBodyElement) innerRef={{...}} content=\\"content2\\" selected={false} focused={false} onClick={[Function: handleSelect]} jssContext={{...}} theme={{...}}>
          <DropdownBodyElement content=\\"content2\\" selected={false} focused={false} onClick={[Function: handleSelect]} classes={{...}}>
            <div className=\\"DropdownBodyElement-root-0-2-8\\" onClick={[Function: handleSelect]}>
              content2
            </div>
          </DropdownBodyElement>
        </WithStyles(DropdownBodyElement)>
      </JssContextSubscriber>
      <JssContextSubscriber isEmpty={false} isFilteredEmpty={false}>
        <WithStyles(DropdownBodyEmpty) innerRef={{...}} isEmpty={false} isFilteredEmpty={false} jssContext={{...}} theme={{...}}>
          <DropdownBodyEmpty isEmpty={false} isFilteredEmpty={false} classes={{...}} />
        </WithStyles(DropdownBodyEmpty)>
      </JssContextSubscriber>
    </div>
  </div>
</DropdownBody>"
`;
